### DISCUSSION TITLE
Question about Aves and XMP metadata handling - why not ExifTool?

### DISCUSSION THREAD

**Discussion by ghost** [deckerst/aves#230] [Q&A] [answered]
Hello there! Before I start, I want to express my gratitude for this app - I’ve yet to find any app that has the functionality as Aves on any mobile platform. There’s nothing like it on iOS, and the only Android alternatives I could find are closed-source (😬) and paid

Some of the more active feature requests on this project and my personal pet peeves are about expanded support for metadata editing (Date, Location, Tags) - some examples of unsupported formats are HEIC and all types of videos. From my understanding by reading the issue posts here, the reason why Aves Gallery can’t add support for these formats is because the [library](https://github.com/dragon66/pixymeta-android) that is used doesn’t support the formats and it also appears to be unmaintained.

Yet, there appears to be other open-source libraries that are much more actively maintained and also supports a much more wider range of image and video formats - namely ExifTool. They have support for essentially every type of video and image format out there, unlike pixymeta-android.

Thus, I’m curious as to why you’ve chosen to use pixymeta-android for this project, which is rather unmaintained and hinders the file formats that Aves Gallery supports, as opposed to ExifTool (and perhaps, other tools like it). My money is on ExifTool not being native to Java/Android platform, but there appears to be other [apps](https://play.google.com/store/apps/details?id=com.exiftool.free&hl=en_US&gl=US) which has successfully implemented ExifTool, so it seems achievable.

Thank you!

**Comment by deckerst** [comment, marked as answer]
Hi! Fair question, and ExifTool is indeed the best tool out there.

I've documented (mostly for myself) the libraries I've considered so far [here](https://github.com/deckerst/aves/wiki/Package-Study#exif=). The main issue with ExifTool is that it's in Perl, which means Aves would have to include the ExifTool library itself, but also a Perl interpreter with it! Also, I could be wrong, but launching Perl processes on Android is likely to have poor performance compared to some Java/C library. And getting Perl to Android in the first place can be done for sure, but it's not trivial and I'm not eager to spend time debugging compilation issues.

Aves currently edits metadata with a combination of [ExifInterface](https://developer.android.com/jetpack/androidx/releases/exifinterface) (which is imperfect, but convenient and maintained) and a [forked version of pixymeta-android](https://github.com/deckerst/pixymeta-android/) (the original being unmaintained, as you noticed).

My goal is not to support all formats out there, and I will not add 50 MB of dependencies for that purpose. In my *opinion*, editing support for JPEG, GIF, PNG and MP4 are a __must__, all the rest is a bonus. For videos, maybe I can leverage FFmpeg (which Aves already uses for playback), but I'd need to dedicate time to look into that. For HEIC, maybe the [HeifWriter](https://developer.android.com/reference/androidx/heifwriter/HeifWriter.html) bundled with Android will do.
